src/pybamm/config.py

```diff
@@ -0,0 +1,141 @@
+import uuid
+import os
+import platformdirs
+from pathlib import Path
+import pybamm
+from inputimeout import inputimeout, TimeoutOccurred
+
+
+def is_running_tests():  # pragma: no cover
+    """
+    Detect if the code is being run as part of a test suite or building docs with Sphinx.
+
+    Returns:
+        bool: True if running tests or building docs, False otherwise.
+    """
+    import sys
+
+    # Check if pytest or unittest is running
+    if any(
+        test_module in sys.modules for test_module in ["pytest", "unittest", "nose"]
+    ):
+        return True
+
+    # Check for GitHub Actions environment variable
+    if "GITHUB_ACTIONS" in os.environ:
+        return True
+
+    # Check for other common CI environment variables
+    ci_env_vars = ["CI", "TRAVIS", "CIRCLECI", "JENKINS_URL", "GITLAB_CI"]
+    if any(var in os.environ for var in ci_env_vars):
+        return True
+
+    # Check for common test runner names in command-line arguments
+    test_runners = ["pytest", "unittest", "nose", "trial", "nox", "tox"]
+    if any(runner in sys.argv[0].lower() for runner in test_runners):
+        return True
+
+    # Check if building docs with Sphinx
+    if "sphinx" in sys.modules:
+        return True
+
+    return False
+
+
+def ask_user_opt_in(timeout=10):
+    """
+    Ask the user if they want to opt in to telemetry.
+
+    Parameters
+    ----------
+    timeout: float, optional
+        The timeout for the user to respond to the prompt. Default is 20 seconds.
+
+    Returns
+    -------
+    bool
+        True if the user opts in, False otherwise.
+    """
+    print(
+        "PyBaMM can collect usage data and send it to the PyBaMM team to "
+        "help us improve the software.\n"
+        "We do not collect any sensitive information such as models, parameters, "
+        "or simulation results - only information on which parts of the code are "
+        "being used and how frequently.\n"
+        "This is entirely optional and does not impact the functionality of PyBaMM.\n"
+        "For more information, see https://docs.pybamm.org/en/latest/source/user_guide/index.html#telemetry"
+    )
+
+    while True:
+        try:
+            user_input = (
+                inputimeout(
+                    prompt="Do you want to enable telemetry? (Y/n): ", timeout=timeout
+                )
+                .strip()
+                .lower()
+            )
+        except TimeoutOccurred:
+            print("\nTimeout reached. Defaulting to not enabling telemetry.")
+            return False
+
+        if user_input in ["yes", "y", ""]:
+            return True
+        elif user_input in ["no", "n"]:
+            return False
+        else:
+            print("\nInvalid input. Please enter 'yes'/'y' or 'no'/'n'.")
```

Comment:
If yes, y, and no response are already yes, then we can probably just assume that anything else is no
```suggestion
        else:
            return False
```

Reply:
I would prefer to avoid the situation where people do want to enable telemetry but disable it by accident